### ISSUE TITLE
Form submissions redirects broken in proxied remix app

### ISSUE THREAD

**Issue by vanceism7** [Shopify/shopify-app-js#1073]
# Issue summary

Before opening this issue, I have:

- [x] Upgraded to the latest version of the relevant packages
  - `@shopify/*` package and version:
  - Node version: v20.11.0
  - Operating system: Windows 10
- [ ] Set `{ logger: { level: LogSeverity.Debug } }` in my configuration, when applicable
- [x] Found a reliable way to reproduce the problem that indicates it's a problem with the package
- [x] Looked for similar issues in this repository

---

I'm working on a shopify remix app, which is using the shopify proxy on the store front side of things to add some additional functionality. In my work, I think I've discovered that form submission actions seem to be broken because of the proxy url rewrites. 

Here's my minimal reproduction

```typescript
const FooPage = () => {
  return (
    <main>
      // You can also use plain `Form` here, the result is the same
      <AppProxyForm action="/apps/proxy/foo" method="post">
        <button type="submit">Click me</button>
      </AppProxyForm>
    </main>
  );
};

export default FooPage;

export async function action({ request }: ActionFunctionArgs) {
  return null;
  // or return redirect("/")
  // Really you can return anything here, the result is the same
}
```

## Expected behavior

- If I'm returning `null`, I expect the page to remain unchanged
- If I'm returning `redirect("/")` - I expect the page to go back to my store front url, e.g: `foo.myshopify.com`
- If I return `redirect("/apps/proxy/foo")` - I expect the page to redirect to `foo.myshopify.com/apps/proxy/foo` - i.e: the url should be unchanged

## Actual behavior
In all of the above expected scenarios, instead of the page remaining unchanged or redirecting properly, the store url always gets rewritten to the url of my actual app.

E.g: (If my app url is `myapp.com`)

- returning `null` redirects me to `myapp.com/apps/proxy/foo` (the url is unchanged *except* `foo.myshopify.com` becomes `myapp.com`)
- returning `redirect("/")` redirects me to `myapp.com` instead of `foo.myshopify.com`
- returning `redirect("/apps/proxy/foo")` redirects me to `myapp.com/apps/proxy/foo` instead of `foo.myshopify.com/apps/proxy/foo`

### Other notes
This issue seems to specifically happen with form submission actions.

For example, the following loader function works as expected

```typescript
export const loader: LoaderFunction = async () => {
  return redirect("/")
};
```

This function will successfully redirect me to `foo.myshopify.com`.

Also, if I call the action function manually using a `fetch` call, this also gets back the response fine without any redirection.

But any call within the context of a form submission redirects away from the shopify store over to my app's direct url.  This is specifically a big issue when using libs like `conform`, which use the form action to run validation on the server side, causing the form to break immediately on tabbing from one form input to another

**Comment by matteodepalo** [comment]
Hi @vanceism7, thank you for opening this issue, the team will take a look.

**Comment by anasdevv** [comment]
Any progress with the findings? This really gives me headaches..

**Comment by lizkenyon** [comment]
Hi there 👋 

Sorry for the delay on this. A couple of things

> seem to be broken because of the proxy url rewrites
Yes, this is a known issue with Remix unfortunately. At this time there are a couple of options. 

1.  Ensure that the URL you are are on for your app proxy page ends in a trailing slash. Like described [here](https://shopify.dev/docs/api/shopify-app-remix/v3/entrypoints/appproxyprovider#:~:text=Because%20Remix%20doesn%27t%20support%20URL%20rewriting%2C%20any%20route%20using%20this%20component%20should%20match%20the%20pathname%20of%20the%20proxy%20URL%20exactly%2C%20and%20end%20in%20a%20trailing%20slash).

2. If you don't need to use the resulting data, and just want to kick off a form submission you can use the [navigate prop](https://remix.run/docs/en/main/components/form#navigate) on the Remix from the Remix form component.

```
 return (
      <AppProxyProvider appUrl={appUrl}>
        <AppProxyForm action="/apps/appProxy" navigate={false} method="post">
          <input type="text" name="field" defaultValue="Enter text here" />
  
          <input type="submit" name="Submit 2" />
        </AppProxyForm>
      </AppProxyProvider>
    );
```
> returning redirect("/") redirects me to myapp.com instead of foo.myshopify.com
Yes, because you are using the Remix redirect it is redirecting based on your app URL. 

I will create a ticket for us to explore adding a redirect helper to the App Proxy context (similar to what we do for the[ Admin context](https://shopify.dev/docs/api/shopify-app-remix/v3/authenticate/admin#redirect-redirecting-to-an-app-route)), that could allow for specifying redirects how you describe in the expected. 

For the time being if you want to redirect to a different URL on the shop, you will need to specify the entire path.

I will be making some updates to the documentation with some more of this context.

**Comment by anasdevv** [comment]
> ```
>  return (
>       <AppProxyProvider appUrl={appUrl}>
>         <AppProxyForm action="/apps/appProxy" navigate={false} method="post">
>           <input type="text" name="field" defaultValue="Enter text here" />
>   
>           <input type="submit" name="Submit 2" />
>         </AppProxyForm>
>       </AppProxyProvider>
>     );
> ```

One more thing i'd like to discuss , As clearly mentioned in the docs , merchants can change the url this means the form will break. Is there any way to get the proxy URL?

**Comment by lizkenyon** [comment]
I agree this is something that there should be a better experience. If you create a new question in the[ new developer community](https://community.shopify.dev/) I can try to bump that with the team that owns App Proxy.
